### PR TITLE
cannon: Fix 64-bit memory access

### LINKED ISSUE #12244: Fix memory.go for the 64-bit VM
`Memory` is currently broken on cannon64.

### PR DESCRIPTION
**Pull request by Inphi**
This patch fully implements 64-bit memory trie operations for the 64-bit VM.

`memory.GetMemory` is replaced with `memory.GetUint32` for clarity. `GetUint32` is useful to have even on cannon64 for fixed 32-bit memory access.

## Testing

Introduced a memory64_test.go that only runs with `-tags=cannon64`. The cannon64 build is not yet enabled in CI and will be fixed as part of https://github.com/ethereum-optimism/optimism/issues/12252. Till then, one can manually test 64-bit memory using:
```bash
go test -tags=cannon64 ./mipsevm/memory/
```

fixes #12244

### CODE REVIEW

File: cannon/mipsevm/memory/memory.go [hunk 1]
```diff
@@ -98,8 +98,9 @@ func (m *Memory) invalidate(addr Word) {
 		return
 	}
 
-	// find the gindex of the first page covering the address
-	gindex := ((uint64(1) << 32) | uint64(addr)) >> PageAddrSize
+	// find the gindex of the first page covering the address: i.e. ((1 << WordSize) | addr) >> PageAddrSize
+	// Avoid 64-bit overflow by expanding the rsh over the OR.
```
Comment (mbaxter) [inline cannon/mipsevm/memory/memory.go]: (nit) Had to stare at this comment a bit - maybe update to:
```suggestion
	// Avoid 64-bit overflow by distributing the right shift across the OR.
```

File: cannon/mipsevm/memory/memory.go [hunk 2]
```diff
@@ -192,7 +193,7 @@ func (m *Memory) pageLookup(pageIndex Word) (*CachedPage, bool) {
 }
 
 func (m *Memory) SetMemory(addr Word, v uint32) {
```
Comment (mbaxter) [inline cannon/mipsevm/memory/memory.go]:
```suggestion
func (m *Memory) SetUint32(addr Word, v uint32) {
```

File: cannon/mipsevm/memory/memory.go [hunk 3]
```diff
@@ -245,7 +246,7 @@ func (m *Memory) GetMemory(addr Word) uint32 {
 }
 
 // GetWord reads the maximum sized value, [arch.Word], located at the specified address.
-// Note: Also known by the MIPS64 specification as a "double-word" memory access.
+// Note: Also referred to by the MIPS64 specification as a "double-word" memory access.
```
Comment (mbaxter) [inline cannon/mipsevm/memory/memory.go:249]: (nit) this is only true for 64-bit right?:
```suggestion
// Note: In 64-bit architectures, corresponds to the MIPS64 specification's "doubleword" memory access.
```

**Review by mbaxter — DISMISSED** [review]
Looks good!  Just left a few language nits you can take or leave.

**Comment by Inphi** [comment]
@mbaxter Could I get your approval again for contracts-bedrock? The memory SetUint32 refactor touches the go-ffi differential-testing script

**Review by mbaxter — APPROVED** [review]

**Pull request merged**